File: rust/parity/specs/rbac_permission_mapper.py
"""Parity spec for PiRBACPermissionMapper.

A spec declares:
  RUST_NAME       - registry key passed to pi_core.run_agent
  run_python(data)-> dict  - run the ORIGINAL python agent, return model_dump()
  SAMPLES         - list of {"input": {...}, "env": {...optional...}}
"""
from _util import load_py_agent

RUST_NAME = "PiRBACPermissionMapper"

_mod = load_py_agent("pi_rbac_permission_mapper.py")


def run_python(data: dict) -> dict:
    agent = _mod.PiRBACPermissionMapper()
    out = agent.map_rbac_permissions(_mod.RBACInput(**data))
    return out.model_dump()


# Clean, least-privilege policy: no wildcards, no escalation.
_CLEAN = (
    '{"Effect": "Allow", "Action": "s3:GetObject", '
    '"Resource": "arn:aws:s3:::my-bucket/*"}'
)
# Wildcard action (JSON style).
_WILDCARD_ACTION = '{"Effect": "Allow", "Action": "*", "Resource": "arn:aws:s3:::b"}'
# Wildcard action (YAML/lower style).
_WILDCARD_ACTION_YAML = "action: '*'\nresource: arn:aws:s3:::b"
# Wildcard resource WITH an Allow effect -> flagged at 70.
_WILDCARD_RESOURCE = '{"Effect": "Allow", "Action": "s3:GetObject", "Resource": "*"}'
# Wildcard resource WITHOUT any Allow effect -> NOT flagged.
_WILDCARD_RESOURCE_NO_ALLOW = '{"Action": "s3:GetObject", "Resource": "*"}'
# Privilege escalation via iam:PassRole.
_PRIV_ESC = '{"Effect": "Allow", "Action": "iam:PassRole", "Resource": "*"}'
# Everything at once: wildcard action + wildcard resource(allow) + escalation.
_ALL_THREE = (
    '{"Effect": "Allow", "Action": "*", "Resource": "*", '
    '"Note": "iam:AttachRolePolicy"}'
)

SAMPLES = [
    {"input": {"policy_file_path": "p.json", "policy_content": _CLEAN}},
    {"input": {"policy_file_path": "p.json", "policy_content": _WILDCARD_ACTION}},
    {"input": {"policy_file_path": "p.yaml", "policy_content": _WILDCARD_ACTION_YAML}},
    {"input": {"policy_file_path": "p.json", "policy_content": _WILDCARD_RESOURCE}},
    {"input": {"policy_file_path": "p.json", "policy_content": _WILDCARD_RESOURCE_NO_ALLOW}},
    {"input": {"policy_file_path": "p.json", "policy_content": _PRIV_ESC}},
    {"input": {"policy_file_path": "p.json", "policy_content": _ALL_THREE}},
    {"input": {"policy_file_path": "p.json", "policy_content": ""}},
    # non-strict env -> WARN path: risk>30 but is_secure stays True.
    {"input": {"policy_file_path": "p.json", "policy_content": _WILDCARD_ACTION},
     "env": {"PI_RBAC_STRICT_MODE": "false"}},
    # explicit strict env -> OVERLY_PERMISSIVE.
    {"input": {"policy_file_path": "p.json", "policy_content": _WILDCARD_ACTION},
     "env": {"PI_RBAC_STRICT_MODE": "true"}},
]

File: rust/crates/pi-agents/src/agents/rbac_permission_mapper.rs
//! Port of `pi_micro_agents/pi_rbac_permission_mapper.py`.
//!
//! Maps IAM/RBAC policies to detect least-privilege violations, wildcard
//! actions/resources, and privilege-escalation risks. Behaviour is a
//! line-for-line mirror of the Python original.

use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
pub struct Input {
    pub policy_file_path: String,
    pub policy_content: String,
}

#[derive(Debug, Serialize, PartialEq)]
pub struct Output {
    pub is_secure: bool,
    pub excessive_permissions: Vec<String>,
    pub risk_score: f64,
    pub status: String,
}

/// Mirrors `is_strict_mode()`: strict unless the env var is set, in which case
/// strictness is determined by a case-insensitive comparison to "true".
fn is_strict_mode() -> bool {
    match std::env::var("PI_RBAC_STRICT_MODE") {
        Ok(v) => v.to_lowercase() == "true",
        Err(_) => true,
    }
}

pub fn map_rbac_permissions(input: &Input) -> Output {
    let content = &input.policy_content;
    let mut excessive: Vec<String> = Vec::new();
    let mut risk_score: f64 = 0.0;

    // Action: * checks
    if content.contains("\"Action\": \"*\"")
        || content.contains("\"action\": \"*\"")
        || content.contains("Action: '*'")
        || content.contains("action: '*'")
    {
        excessive.push(
            "Wildcard Action: Policy allows arbitrary actions ('*') which violates least-privilege principles.".to_string(),
        );
        risk_score = risk_score.max(95.0);
    }

    // Resource: * checks
    if content.contains("\"Resource\": \"*\"")
        || content.contains("\"resource\": \"*\"")
        || content.contains("Resource: '*'")
        || content.contains("resource: '*'")
    {
        if content.contains("Effect: Allow")
            || content.contains("\"Effect\": \"Allow\"")
            || content.contains("\"effect\": \"allow\"")
        {
            excessive.push(
                "Wildcard Resource: Policy allows actions on all target resources which may cause data leakage.".to_string(),
            );
            risk_score = risk_score.max(70.0);
        }
    }

    // Privilege escalation checks: iam:PassRole or AttachRolePolicy
    if content.contains("iam:PassRole")
        || content.contains("iam:AttachRolePolicy")
        || content.contains("iam:PutUserPolicy")
    {
        excessive.push(
            "Privilege Escalation Risk: Permission grants critical IAM management controls (e.g. iam:PassRole).".to_string(),
        );
        risk_score = risk_score.max(90.0);
    }

    let mut is_sec = true;
    if risk_score > 30.0 && is_strict_mode() {
        is_sec = false;
    }

    let mut status = if is_sec { "PASSED" } else { "OVERLY_PERMISSIVE" }.to_string();
    if risk_score > 0.0 && is_sec {
        status = "WARN_PERMISSIVE".to_string();
    }

    Output {
        is_secure: is_sec,
        excessive_permissions: excessive,
        risk_score,
        status,
    }
}

pub fn run_json(input_json: &str) -> Result<String, String> {
    let input: Input = serde_json::from_str(input_json).map_err(|e| e.to_string())?;
    let out = map_rbac_permissions(&input);
    serde_json::to_string(&out).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run(content: &str) -> Output {
        map_rbac_permissions(&Input {
            policy_file_path: "policy.json".into(),
            policy_content: content.into(),
        })
    }

    #[test]
    fn clean_policy_passes() {
        let o = run("{\"Action\": \"s3:GetObject\", \"Resource\": \"arn:aws:s3:::bucket\"}");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);
        assert!(o.excessive_permissions.is_empty());
    }

    #[test]
    fn wildcard_action_flagged() {
        let o = run("{\"Action\": \"*\"}");
        assert!(!o.is_secure);
        assert_eq!(o.status, "OVERLY_PERMISSIVE");
        assert_eq!(o.risk_score, 95.0);
        assert_eq!(o.excessive_permissions.len(), 1);
    }

    #[test]
    fn wildcard_resource_needs_allow_effect() {
        // Resource:* without an Allow effect must NOT be flagged.
        let o = run("{\"Resource\": \"*\"}");
        assert!(o.is_secure);
        assert_eq!(o.status, "PASSED");
        assert_eq!(o.risk_score, 0.0);

        // With an Allow effect it is flagged at 70.0.
        let o2 = run("{\"Effect\": \"Allow\", \"Resource\": \"*\"}");
        assert!(!o2.is_secure);
        assert_eq!(o2.risk_score, 70.0);
        assert_eq!(o2.status, "OVERLY_PERMISSIVE");
    }

    #[test]
    fn privilege_escalation_flagged() {
        let o = run("{\"Action\": \"iam:PassRole\"}");
        assert!(!o.is_secure);
        assert_eq!(o.risk_score, 90.0);
    }
}
